File: internal/webui/testdata/pwa.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { runInNewContext } from "node:vm";
import { webcrypto } from "node:crypto";

const asset = (name) => new URL(`../static/offline/${name}`, import.meta.url);

test("offline manifest is stable and scoped", async () => {
  const manifest = JSON.parse(await readFile(asset("manifest.json"), "utf8"));
  assert.equal(manifest.id, undefined);
  assert.equal(manifest.start_url, "./");
  assert.equal(manifest.scope, "./");
  assert.equal(manifest.display, "standalone");
  assert.deepEqual(manifest.icons.map((icon) => icon.src), ["./icon.svg", "./icon-512.svg"]);
});

test("offline app identities inherit deployment-specific start URLs", async () => {
  const manifest = JSON.parse(await readFile(asset("manifest.json"), "utf8"));
  const identities = ["a", "b"].map(prefix => {
    const start = new URL(manifest.start_url, `https://reader.example/${prefix}/ui/offline/manifest.json`);
    return manifest.id === undefined ? start.href : new URL(manifest.id, start.origin).href;
  });
  assert.deepEqual(identities, ["https://reader.example/a/ui/offline/", "https://reader.example/b/ui/offline/"]);
});

test("offline worker names only the shell assets", async () => {
  const worker = await readFile(asset("sw.js"), "utf8");
  for (const name of ["./shell.html", "./manifest.json", "./offline.css", "./offline.js", "./offline-shelf.js", "./icon.svg", "./icon-512.svg", "./apple-touch-icon.png"]) {
    assert.match(worker, new RegExp(name.replace(".", "\\.")));
  }
  for (const forbidden of ["/v1/", "Authorization", "token", "book"]) {
    assert.doesNotMatch(worker.toLowerCase(), new RegExp(forbidden.toLowerCase()));
  }
  for (const asset of ["./read/", "./assets/offline-account.js", "./assets/offline-storage.js", "./assets/offline-sync.js", "./assets/reader-app.js", "./assets/vendor/readium/readium.js"]) {
    assert.match(worker, new RegExp(asset.replaceAll(".", "\\.")));
  }
  assert.doesNotMatch(worker, /cache\.put/);
});

test("offline shell contains no personalized placeholders", async () => {
  const shell = await readFile(asset("shell.html"), "utf8");
  assert.doesNotMatch(shell, /alice|csrf|token|api|book id/i);
  assert.match(shell, /Offline shelf/);
  assert.match(shell, /offline\.js/);
  assert.match(shell, /offline-shelf\.js/);
  assert.match(shell, /\.\.\/library/);
});

async function workerHarness(root = "/sync/ui/offline/", revision = "new") {
  const listeners = new Map();
  const stores = new Map();
  const deleted = [];
  const prefix = "liseur-sync-offline-shell-" + encodeURIComponent(root) + "-";
  const current = prefix + revision;
  let installation = [];
  let claimed = false;
  const cacheFor = name => {
    if (!stores.has(name)) stores.set(name, new Map());
    return {
      match: async request => stores.get(name).get(request.url || request)?.clone(),
      addAll: async requests => { installation = requests; },
    };
  };
  runInNewContext((await readFile(asset("sw.js"), "utf8"))
    .replaceAll("__OFFLINE_SHELL_REVISION__", revision), {
    self: {
      location: new URL("https://reader.example" + root + "sw.js"),
      addEventListener: (name, handler) => listeners.set(name, handler),
      clients: { claim: async () => { claimed = true; } },
    },
    caches: {
      open: async name => cacheFor(name),
      keys: async () => [...stores.keys()],
      delete: async name => { deleted.push(name); return stores.delete(name); },
    },
    URL, Request, Response, Headers, crypto: webcrypto, Uint8Array, btoa,
  });
  return {
    stores, deleted, prefix, current,
    get installation() { return installation; },
    get claimed() { return claimed; },
    put(name, path, response) {
      cacheFor(name);
      stores.get(name).set("https://reader.example" + path, response);
    },
    async lifecycle(name) {
      let promise;
      listeners.get(name)({ waitUntil: value => { promise = value; } });
      await promise;
    },
    fetch(path, mode = "navigate", method = "GET") {
      let promise;
      listeners.get("fetch")({
        request: { url: "https://reader.example" + path, mode, method },
        respondWith: value => { promise = value; },
      });
      return promise;
    },
  };
}

test("worker installs fresh assets without forcing an active reader to upgrade", async () => {
  const worker = await workerHarness();
  await worker.lifecycle("install");
  assert.ok(worker.installation.length > 20);
  for (const request of worker.installation) {
    assert.equal(request.cache, "reload");
    assert.ok(new URL(request.url).pathname.startsWith("/sync/ui/offline/"));
  }
  assert.equal(worker.claimed, false);
});

test("worker activation cleans only older generations of its deployment", async () => {
  const worker = await workerHarness();
  worker.put(worker.prefix + "old", "/sync/ui/offline/shell.html", new Response("old"));
  worker.put(worker.current, "/sync/ui/offline/shell.html", new Response("new"));
  worker.put("liseur-sync-offline-shell-%2Fother%2Fui%2Foffline%2F-old", "/other/ui/offline/shell.html", new Response("other"));
  await worker.lifecycle("activate");
  assert.deepEqual(worker.deleted, [worker.prefix + "old"]);
  assert.equal(worker.claimed, true);
  assert.equal(worker.stores.size, 2);
});

test("worker serves reader and modules from the same generation with fresh CSP nonces", async () => {
  const worker = await workerHarness();
  const path = "/sync/ui/offline/read/";
  const response = new Response('<script type="module" nonce="original" src="../assets/reader-app.js"></script>', {
    headers: {
      "Content-Security-Policy": "default-src 'none'; script-src 'nonce-original' 'strict-dynamic'",
      "Content-Type": "text/html",
      "Content-Length": "123",
    },
  });
  worker.put(worker.current, path, response);
  worker.put(worker.prefix + "old", "/sync/ui/offline/assets/reader-app.js", new Response("old"));
  worker.put(worker.current, "/sync/ui/offline/assets/reader-app.js", new Response("current"));
  const first = await worker.fetch(path + "?book=one");
  const second = await worker.fetch(path + "?book=two");
  const nonce = first.headers.get("Content-Security-Policy").match(/'nonce-([^']+)'/)[1];
  assert.notEqual(nonce, "original");
  assert.match(await first.text(), new RegExp(`nonce="${nonce.replace(/[+]/g, "\\+")}"`));
  assert.notEqual(first.headers.get("Content-Security-Policy"), second.headers.get("Content-Security-Policy"));
  assert.equal(first.headers.get("Content-Length"), null);
  assert.equal(await (await worker.fetch("/sync/ui/offline/assets/reader-app.js", "cors")).text(), "current");
});

test("worker leaves API, account bootstrap, unrelated navigation and mutations alone", async () => {
  const worker = await workerHarness();
  for (const path of ["/sync/v1/ops", "/sync/ui/offline/account", "/sync/ui/login", "/other/ui/offline/"]) {
    assert.equal(worker.fetch(path), undefined);
  }
  assert.equal(worker.fetch("/sync/ui/offline/shell.html", "cors", "POST"), undefined);
  const response = await worker.fetch("/sync/ui/offline/");
  assert.equal(response.type, "error");
});

File: internal/webui/pwa_test.go
package webui

import (
	"bytes"
	"encoding/json"
	"image/png"
	"io"
	"io/fs"
	"net/http"
	"strings"
	"testing"
	"testing/fstest"

	"github.com/chmouel/liseur-sync/internal/config"
)

func TestOfflineShellRequiresSessionOnline(t *testing.T) {
	ts, _ := testServer(t)

	resp, err := noRedirect().Get(ts.URL + "/ui/offline/")
	if err != nil {
		t.Fatal(err)
	}
	resp.Body.Close()
	if resp.StatusCode != http.StatusSeeOther || resp.Header.Get("Location") != "../login" {
		t.Fatalf("unauthenticated shell: got %d -> %q", resp.StatusCode, resp.Header.Get("Location"))
	}

	cookie := loginCookie(t, ts)
	req, err := http.NewRequest(http.MethodGet, ts.URL+"/ui/offline/", nil)
	if err != nil {
		t.Fatal(err)
	}
	req.AddCookie(cookie)
	resp, err = http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	bodyBytes, _ := io.ReadAll(resp.Body)
	resp.Body.Close()
	code, body := resp.StatusCode, string(bodyBytes)
	if code != http.StatusOK {
		t.Fatalf("authenticated shell: got %d", code)
	}
	if got := resp.Header.Get("Content-Security-Policy"); got != uiPolicy {
		t.Fatalf("offline shell CSP changed: %q", got)
	}
	for _, leaked := range []string{"alice", "hunter2hunter", `name="csrf"`, "v1/"} {
		if strings.Contains(body, leaked) {
			t.Errorf("offline shell leaked %q", leaked)
		}
	}
	if !strings.Contains(body, "Offline shelf") {
		t.Fatal("authenticated shell did not render the generic placeholder")
	}

	req, err = http.NewRequest(http.MethodGet, ts.URL+"/ui/offline/account", nil)
	if err != nil {
		t.Fatal(err)
	}
	req.AddCookie(cookie)
	resp, err = http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	var account struct {
		Account string `json:"account"`
		CSRF    string `json:"csrf"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&account); err != nil {
		t.Fatal(err)
	}
	if resp.StatusCode != http.StatusOK || account.Account != "u1" || account.CSRF == "" {
		t.Fatalf("offline account bootstrap: status=%d account=%q csrf=%t",
			resp.StatusCode, account.Account, account.CSRF != "")
	}
	if got := resp.Header.Get("Cache-Control"); got != "no-store" {
		t.Fatalf("offline account cache policy: %q", got)
	}
}

func TestOfflineAssetsArePublicAndRelative(t *testing.T) {
	ts, _ := testServer(t)

	manifestResp, err := noRedirect().Get(ts.URL + "/ui/offline/manifest.json")
	if err != nil {
		t.Fatal(err)
	}
	defer manifestResp.Body.Close()
	if manifestResp.StatusCode != http.StatusOK {
		t.Fatalf("manifest: got %d", manifestResp.StatusCode)
	}
	var manifest struct {
		ID       string `json:"id"`
		StartURL string `json:"start_url"`
		Scope    string `json:"scope"`
		Display  string `json:"display"`
		Icons    []struct {
			Src     string `json:"src"`
			Sizes   string `json:"sizes"`
			Type    string `json:"type"`
			Purpose string `json:"purpose"`
		} `json:"icons"`
	}
	if err := json.NewDecoder(manifestResp.Body).Decode(&manifest); err != nil {
		t.Fatal(err)
	}
	if manifest.ID != "" || manifest.StartURL != "./" ||
		manifest.Scope != "./" || manifest.Display != "standalone" {
		t.Fatalf("manifest metadata is not stable and scoped: %+v", manifest)
	}
	if len(manifest.Icons) != 5 || manifest.Icons[0].Src != "./icon.svg" ||
		manifest.Icons[1].Src != "./icon-512.svg" {
		t.Fatalf("manifest icon is not relative: %+v", manifest.Icons)
	}
	for _, icon := range manifest.Icons {
		if !strings.HasPrefix(icon.Src, "./") {
			t.Fatalf("icon URL is not relative: %q", icon.Src)
		}
		resp, err := noRedirect().Get(ts.URL + "/ui/offline/" + strings.TrimPrefix(icon.Src, "./"))
		if err != nil {
			t.Fatal(err)
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		if resp.StatusCode != http.StatusOK || resp.Header.Get("Content-Type") != icon.Type {
			t.Fatalf("icon %s: status %d, type %q", icon.Src, resp.StatusCode, resp.Header.Get("Content-Type"))
		}
		if icon.Type == "image/png" {
			im, err := png.Decode(bytes.NewReader(body))
			if err != nil {
				t.Fatalf("icon %s: %v", icon.Src, err)
			}
			want := 512
			if icon.Sizes == "192x192" {
				want = 192
			}
			if im.Bounds().Dx() != want || im.Bounds().Dy() != want {
				t.Errorf("icon %s: dimensions %v do not match %s", icon.Src, im.Bounds(), icon.Sizes)
			}
			if icon.Purpose == "maskable" {
				for y := 0; y < want; y++ {
					for x := 0; x < want; x++ {
						if _, _, _, a := im.At(x, y).RGBA(); a != 0xffff {
							t.Fatalf("maskable icon is transparent at %d,%d", x, y)
						}
					}
				}
			}
		}
	}

	for _, name := range []string{"shell.html", "offline.css", "offline.js", "offline-shelf.js", "icon.svg", "icon-512.svg", "apple-touch-icon.png", "sw.js"} {
		resp, err := noRedirect().Get(ts.URL + "/ui/offline/" + name)
		if err != nil {
			t.Fatal(err)
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			t.Errorf("%s: got %d", name, resp.StatusCode)
		}
		if strings.Contains(string(body), "alice") || strings.Contains(string(body), "csrf") {
			t.Errorf("%s contains personalized data", name)
		}
	}
	for _, name := range []string{"reader-app.js", "reader-engine.js", "reader-publication.js", "vendor/readium/readium.js"} {
		resp, err := noRedirect().Get(ts.URL + "/ui/offline/assets/" + name)
		if err != nil {
			t.Fatal(err)
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			t.Errorf("reader asset %s: got %d", name, resp.StatusCode)
		}
		if strings.Contains(string(body), "hunter2hunter") {
			t.Errorf("reader asset %s contains credential material", name)
		}
	}
}

func TestOfflineServiceWorkerIsBoundedToShellAssets(t *testing.T) {
	ts, _ := testServer(t)
	resp, err := noRedirect().Get(ts.URL + "/ui/offline/sw.js")
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	script := string(body)
	for _, required := range []string{
		"liseur-sync-offline-shell-",
		"./shell.html",
		"./manifest.json",
		"./offline.css",
		"./offline.js",
		"./offline-shelf.js",
		"./icon.svg",
		"./icon-512.svg",
		"./icon-192.png",
		"./icon-512.png",
		"./icon-maskable-512.png",
		"request.mode === 'navigate'",
	} {
		if !strings.Contains(script, required) {
			t.Errorf("service worker missing %q", required)
		}
	}
	for _, forbidden := range []string{"/v1/", "Authorization", "token", "book"} {
		if strings.Contains(strings.ToLower(script), strings.ToLower(forbidden)) {
			t.Errorf("service worker contains forbidden %q", forbidden)
		}
	}
	for _, asset := range []string{
		"./read/",
		"./assets/offline-account.js",
		"./assets/offline-storage.js",
		"./assets/reader-app.js",
		"./assets/vendor/readium/readium.js",
	} {
		if !strings.Contains(script, asset) {
			t.Errorf("service worker missing reader asset %q", asset)
		}
	}
	if strings.Contains(script, "cache.put") {
		t.Fatal("service worker must not cache network responses")
	}
	if strings.Contains(script, offlineRevisionPlaceholder) {
		t.Fatal("service worker revision was not rendered")
	}
	if strings.Contains(script, "skipWaiting") {
		t.Fatal("an upgrade must wait for readers using the old shell to close")
	}
}

func TestOfflineShellRevisionTracksEmbeddedAssetsAndReader(t *testing.T) {
	assets := fstest.MapFS{}
	err := fs.WalkDir(staticFS, "static", func(path string, entry fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if entry.IsDir() {
			return nil
		}
		data, err := fs.ReadFile(staticFS, path)
		if err != nil {
			return err
		}
		assets[path] = &fstest.MapFile{Data: data}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	revision := func(reader string) string {
		t.Helper()
		got, err := offlineShellRevision(assets, []byte(reader))
		if err != nil {
			t.Fatal(err)
		}
		return got
	}
	baseline := revision("reader")
	if again := revision("reader"); again != baseline {
		t.Fatal("unchanged assets produced a different worker revision")
	}
	for path := range assets {
		_, isReaderAsset := offlineReaderAssets[strings.TrimPrefix(path, "static/")]
		_, isShellAsset := offlineAssetTypes[strings.TrimPrefix(path, "static/offline/")]
		if !isReaderAsset && !isShellAsset {
			continue
		}
		old := assets[path]
		assets[path] = &fstest.MapFile{Data: append(append([]byte(nil), old.Data...), '\n')}
		if revision("reader") == baseline {
			t.Errorf("changing %s did not update the worker", path)
		}
		assets[path] = old
	}
	if revision("new reader") == baseline {
		t.Fatal("changing the reader shell did not update the worker")
	}
	delete(assets, "static/offline/shell.html")
	if _, err := offlineShellRevision(assets, nil); err == nil {
		t.Fatal("missing shell asset must fail revision generation")
	}
}

func TestUILinksToRelativeOfflineInstallSurface(t *testing.T) {
	ts, _ := testServer(t)
	cookie := loginCookie(t, ts)
	code, body := page(t, ts, cookie, "/ui/library")
	if code != http.StatusOK {
		t.Fatalf("library: got %d", code)
	}
	for _, want := range []string{
		`rel="manifest" href="offline/manifest.json"`,
		`data-pwa-base="offline/"`,
	} {
		if !strings.Contains(body, want) {
			t.Errorf("library missing relative PWA surface %q", want)
		}
	}
}

func TestSeparateReaderOriginDisablesOfflineSurface(t *testing.T) {
	ts, _ := testServerCfg(t, func(cfg *config.Config) {
		cfg.ReaderOrigin = "https://reader.example.com"
	}, nil)

	for _, path := range []string{"/ui/offline/", "/ui/offline/manifest.json"} {
		resp, err := noRedirect().Get(ts.URL + path)
		if err != nil {
			t.Fatal(err)
		}
		resp.Body.Close()
		if resp.StatusCode != http.StatusNotFound {
			t.Errorf("%s: got %d, want 404", path, resp.StatusCode)
		}
	}

	cookie := loginCookie(t, ts)
	code, body := page(t, ts, cookie, "/ui/library")
	if code != http.StatusOK {
		t.Fatalf("library: got %d", code)
	}
	for _, absent := range []string{
		`rel="manifest"`,
		`data-pwa-base="offline/`,
		`offline-account.js`,
		`offline-download.js`,
		`href="offline/"`,
	} {
		if strings.Contains(body, absent) {
			t.Errorf("separate reader origin still exposes %q", absent)
		}
	}
}

func TestOfflineInstallURLsStayUnderTheUIPath(t *testing.T) {
	for path, want := range map[string]string{
		"/ui/":                "offline/",
		"/ui/library":         "offline/",
		"/ui/books/book-id":   "../offline/",
		"/ui/books/book/read": "../../offline/",
	} {
		if got := relPrefix(path) + "offline/"; got != want {
			t.Errorf("%s: offline base %q, want %q", path, got, want)
		}
	}
}
